firestore/integration_test_internal/src/filter_test.cc

```diff
@@ -0,0 +1,472 @@
+  Query query4 = collection.Where(filter4);
+  Query query5 = collection.Where(filter5);
+  Query query6 = collection.Where(filter6);
+
+  EXPECT_EQ(collection, query1);
+  EXPECT_EQ(collection, query2);
+  EXPECT_EQ(collection, query3);
+  EXPECT_EQ(collection, query4);
+  EXPECT_EQ(collection, query5);
+  EXPECT_EQ(collection, query6);
+}
+
+TEST_F(FilterTest, CompositeComparison) {
+  Filter filter1 = Filter::ArrayContains("foo", FieldValue::Integer(42));
+  Filter filter2 = Filter::EqualTo("foo", FieldValue::Integer(42));
+  Filter filter3 = Filter::NotEqualTo("foo", FieldValue::Integer(42));
+  Filter filter4 = Filter::GreaterThan("foo", FieldValue::Integer(42));
+
+  Filter and1 = Filter::And(filter1);
+  Filter and2 = Filter::And(filter1, filter2);
+  Filter and3 = Filter::And(filter1, filter2, filter3);
+  Filter and4 = Filter::And(filter1, filter2, filter3, filter4);
+
+  Filter or1 = Filter::Or(filter1);
+  Filter or2 = Filter::Or(filter1, filter2);
+  Filter or3 = Filter::Or(filter1, filter2, filter3);
+  Filter or4 = Filter::Or(filter1, filter2, filter3, filter4);
+
+  EXPECT_EQ(and1, and1);
+  EXPECT_EQ(and2, and2);
+  EXPECT_EQ(and3, and3);
+  EXPECT_EQ(and4, and4);
+
+  EXPECT_EQ(or1, or1);
+  EXPECT_EQ(or2, or2);
+  EXPECT_EQ(or3, or3);
+  EXPECT_EQ(or4, or4);
+
+  // Is equal because single filter composite is same as filter itself.
+  EXPECT_EQ(and1, or1);
+
+  EXPECT_NE(and2, or2);
+  EXPECT_NE(and3, or3);
+  EXPECT_NE(and4, or4);
+
+  EXPECT_NE(and1, and2);
+  EXPECT_NE(and1, and3);
+  EXPECT_NE(and1, and4);
+  EXPECT_NE(and2, and3);
+  EXPECT_NE(and2, and4);
+  EXPECT_NE(and3, and4);
+
+  EXPECT_NE(or1, or2);
+  EXPECT_NE(or1, or3);
+  EXPECT_NE(or1, or4);
+  EXPECT_NE(or2, or3);
+  EXPECT_NE(or2, or4);
+  EXPECT_NE(or3, or4);
+}
+
+TEST_F(FilterTest, QueryWhereComposite) {
+  MapFieldValue doc_aa = {{"x", FieldValue::String("a")},
+                          {"y", FieldValue::String("a")}};
+  MapFieldValue doc_ab = {{"x", FieldValue::String("a")},
+                          {"y", FieldValue::String("b")}};
+  MapFieldValue doc_ba = {{"x", FieldValue::String("b")},
+                          {"y", FieldValue::String("a")}};
+  MapFieldValue doc_bb = {{"x", FieldValue::String("b")},
+                          {"y", FieldValue::String("b")}};
+  CollectionReference collection = Collection(
+      {{"aa", doc_aa}, {"ab", doc_ab}, {"ba", doc_ba}, {"bb", doc_bb}});
+
+  Filter filter_xa = Filter::EqualTo("x", FieldValue::String("a"));
+  Filter filter_yb = Filter::EqualTo("y", FieldValue::String("b"));
+
+  QuerySnapshot snapshot1 =
+      ReadDocuments(collection.Where(Filter::And(filter_xa, filter_yb)));
+  EXPECT_EQ(std::vector<MapFieldValue>({doc_ab}),
+            QuerySnapshotToValues(snapshot1));
+
+  QuerySnapshot snapshot2 =
+      ReadDocuments(collection.Where(Filter::Or(filter_xa, filter_yb)));
+  EXPECT_EQ(std::vector<MapFieldValue>({doc_aa, doc_ab, doc_bb}),
+            QuerySnapshotToValues(snapshot2));
+}
```

Comment:
Can you please add a few more tests?

1- have nested composite filters
* `col.where(and(A, or(B, and(C, D))))`

2- have empty composite filters
* `query.where(and())` result set should be same as `query` result set
* `query.where(or())` result set should be same as `query` result set

3- Have composite filter with one filter inside it
* `query.where(and(A))` should have the same result set as `query.where(A)`
* `query.where(or(A))` should have the same result set as `query.where(A)`

4- mix of 1 and 2 and 3
* `query.where(and(or(and(A)),or(and(or())))` should have the same result set as `query.where(A)`
* `query.where(or(and(or(A)),and(or(and())))` should have the same result set as `query.where(A)`

There's some logic implemented in the C++ layer for these scenarios and so we should test them here.